chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,832 @@
+//! `store.flush_to_permanent_store(block 2 id)`, then `Block 1` and `Block 2`
+//! will be moved to the permanent storage. If you call
+//! `store.flush_to_permanent_store(block 3 id)`, `Block 3` will also be moved
+//! to the permanent store, but `Block 3'` will still exist. Note that if you
+//! call `store.get_blocks_by_chain_length(3)` only `Block 3` (which is in the
+//! permanent store) will be returned; and you cannot call
+//! `store.flush_to_permanent_store(block 3' id)` now.
+//!
+//! __fig.1 - note that root does not actually exist, this is an ID referredby__
+//! __the first block in the chain__
+//!
+//! ```ignore
+//! +---------+       +---------+
+//! |         |       |         |
+//! | Block 4 |       | Block 4'|
+//! |         |       |         |
+//! +---------+       +---------+
+//!      |                 |
+//!      |                 |
+//!      v                 v
+//! +---------+       +---------+
+//! |         |       |         |
+//! | Block 3 |       | Block 3'|
+//! |         |       |         |
+//! +---------+       +---------+
+//!      |                 |
+//!      |                 |
+//!      v                 |
+//! +---------+            |
+//! |         |            |
+//! | Block 2 +<-----------+
+//! |         |
+//! +---------+
+//!      |
+//!      |
+//!      v
+//! +---------+
+//! |         |
+//! | Block 1 |
+//! |         |
+//! +---------+
+//!      |
+//!      |
+//!      v
+//! +---------+
+//! |         |
+//! |  (root) |
+//! |         |
+//! +---------+
+//! ```
+//!
+//! ## Removing stale branches
+//!
+//! If you want to clean up branches, do the following:
+//!
+//! * Call `store.get_tips_ids()`, in our example it will return
+//!   `[Block 4 id, Block 4' id]`;
+//! * Determine which branches do you want to remove.
+//! * Call, for example, `store.prune_branch(Block 4' id)`.
+//!
+//! ## Performance benefits of permanent storage
+//!
+//! Since blocks in the permanent storage are stored just one after another (the
+//! structure is `block_length.to_le_bytes() ++ block_bytes` and a file with
+//! references to blocks in the order they were added to the storage), it allows
+//! for the following scenarios:
+//!
+//! * Very fast block iteration
+//! * Transferring a portion of the data file over the network without locking
+//!   it.
+//! * O(1) access by chain length.
+//!
+//! __fig. 2 - permanent storage structure__
+//!
+//! ```ignore
+//! store                  block no. index
+//!
+//! +--------------+       +-------------+
+//! |              |       |             |
+//! | len(block1)  |       | block1 pos  |
+//! |              |       |             |
+//! +--------------+       +-------------+
+//! |              |
+//! | block1       |
+//! |              |
+//! +--------------+       +-------------+
+//! |              |       |             |
+//! | len(block2)  |       | block2 pos  |
+//! |              |       |             |
+//! +--------------+       +-------------+
```

Comment:
You don't need to store the `len(block*)` in the store. It is known from the `block no. index`. For example the length of the `block N` is the `(block N+1 pos)    -   (block N pos)`

Reply:
Yes, I plan to remove that.